File: bindings/swift/Tests/GurmukhiUtilsTests/Guut.swift
import Foundation
import XCTest

typealias Fn = (String) -> String

struct TestUnit: Decodable {
    let name: String
    let functions: [String]
    let type: String
    let assertions: Assertions

    enum Assertions: Decodable {
        case list([String])
        case map([String: String])

        init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            if let list = try? container.decode([String].self) {
                self = .list(list)
            } else if let map = try? container.decode([String: String].self) {
                self = .map(map)
            } else {
                throw DecodingError.typeMismatch(
                    Assertions.self,
                    DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Expected array or object")
                )
            }
        }
    }
}

func guut(_ testName: String, functions: [String: Fn], file: StaticString = #file, line: UInt = #line) {
    let url = URL(fileURLWithPath: #filePath)
        .deletingLastPathComponent() // Tests/GurmukhiUtilsTests
        .deletingLastPathComponent() // Tests
        .deletingLastPathComponent() // swift
        .deletingLastPathComponent() // bindings
        .deletingLastPathComponent() // root
        .appendingPathComponent("test/\(testName).json")

    guard let data = try? Data(contentsOf: url) else {
        XCTFail("Could not read \(url.path)", file: file, line: line)
        return
    }

    guard let units = try? JSONDecoder().decode([TestUnit].self, from: data) else {
        XCTFail("Could not decode \(testName).json", file: file, line: line)
        return
    }

    for unit in units {
        for funcName in unit.functions {
            guard let fn = functions[funcName] else {
                XCTFail("Unknown function: \(funcName)", file: file, line: line)
                continue
            }

            switch unit.type {
            case "is":
                if case .map(let assertions) = unit.assertions {
                    for (input, expected) in assertions {
                        XCTAssertEqual(fn(input), expected, "\(funcName)(\(input))")
                    }
                }
            case "is not":
                if case .map(let assertions) = unit.assertions {
                    for (input, expected) in assertions {
                        XCTAssertNotEqual(fn(input), expected, "\(funcName)(\(input))")
                    }
                }
            case "is-self":
                let inputs: [String]
                switch unit.assertions {
                case .list(let list): inputs = list
                case .map(let map): inputs = Array(map.keys)
                }
                for input in inputs {
                    XCTAssertEqual(fn(input), input, "\(funcName)(\(input)) should be self")
                }
            default:
                XCTFail("Unknown test type: \(unit.type)")
            }
        }
    }
}

File: bindings/kotlin/src/test/kotlin/Guut.kt
package gurmukhi_utils.test

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals

typealias Fn = (String) -> String

fun guut(testName: String, functions: Map<String, Fn>) {
    val file = File("../../test/$testName.json")
    val type = object : TypeToken<List<Map<String, Any>>>() {}.type
    val units: List<Map<String, Any>> = Gson().fromJson(file.readText(), type)

    for (unit in units) {
        val unitFunctions = (unit["functions"] as List<*>).filterIsInstance<String>()
        val unitType = unit["type"] as String
        val assertions = unit["assertions"]!!

        for (funcName in unitFunctions) {
            val fn = functions[funcName] ?: error("Unknown function: $funcName")

            when (unitType) {
                "is" -> {
                    for ((input, expected) in assertions as Map<*, *>) {
                        assertEquals(expected as String, fn(input as String), "$funcName($input)")
                    }
                }
                "is not" -> {
                    for ((input, expected) in assertions as Map<*, *>) {
                        assertNotEquals(expected as String, fn(input as String), "$funcName($input)")
                    }
                }
                "is-self" -> {
                    val inputs = when (assertions) {
                        is List<*> -> assertions.filterIsInstance<String>()
                        is Map<*, *> -> assertions.keys.filterIsInstance<String>()
                        else -> error("Unexpected assertions type: ${assertions::class}")
                    }
                    for (input in inputs) {
                        assertEquals(input, fn(input), "$funcName($input) should be self")
                    }
                }
            }
        }
    }
}
